File: 01_Array_and_String/C++ Code/05_345. Reverse Vowels of a String.cpp
// Problem Link -> https://leetcode.com/problems/reverse-vowels-of-a-string/description/

class Solution {
public:
    bool checkIsVowel(char ch) {
        ch = tolower(ch);
        return ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u';
    }

    string reverseVowels(string s) {
        int left = 0;
        int right = s.size() - 1;

        while (left <= right) {
            if (checkIsVowel(s[left]) && checkIsVowel(s[right])) {
                swap(s[left], s[right]);
                left++, right--;
            } else if (!checkIsVowel(s[left])) {
                left++;
            } else {
                right--;
            }
        }

        return s;
    }
};



// Time Complexity ->O(n)
// Space Complexity ->O(1)

File: 01_Array_and_String/Python Code/05_345. Reverse Vowels of a String.py
# Problem Link -> https://leetcode.com/problems/reverse-vowels-of-a-string/description/

class Solution:
    def reverseVowels(self, s: str) -> str:
        s = list(s)
        vowels = set('aeiouAEIOU')
        i, j = 0, len(s) - 1

        while(i < j):
            if(s[i] not in vowels):
                i += 1
            elif (s[j] not in vowels):
                j -= 1
            else:
                s[i],s[j] = s[j],s[i]
                i += 1
                j -= 1
                
        return "".join(s)

      

# Time Complexity ->O(n)
# Space Complexity ->O(1)
